doc/sphinx/proof-engine/ltac2.rst

```diff
@@ -659,168 +734,476 @@ insert in a concise way an Ltac2 variable of type :n:`constr` into a Coq term.
+
+   `multi_match!`
+      If :n:`@ltac2_expr` fails, continue matching with the next branch.
+      When a :n:`@ltac2_expr` succeeds for a branch, subsequent failures
+      (after the `multi_match!`) causing consumption of all the successes
+      of :n:`@ltac2_expr` trigger selection of a new matching branch.
+      :ref:`Example<ltac2_match_vs_multimatch_ex>`.
+
+   :n:`@cpattern`
+      The syntax of :token:`cpattern` is
+      the same as that of :token:`term`\s, but it can contain pattern matching
+      metavariables in the form :n:`?@ident`.  :g:`_` can be used to match
+      irrelevant terms.
+
+      .. todo Example is broken :ref:`Example<ltac2_match_with_holes_ex>`.
+
+      .. todo Didn't understand the following 2 paragraphs well enough to revise
+         see https://github.com/coq/coq/pull/12103#discussion_r436297754 for a
+         possible example
+
+      When a metavariable in the form :n:`?id` occurs under binders,
+      say :n:`x__1, …, x__n` and the expression matches, the
+      metavariable is instantiated by a term which can then be used in any
+      context which also binds the variables :n:`x__1, …, x__n` with
+      same types. This provides with a primitive form of matching under
+      context which does not require manipulating a functional term.
+
+      There is also a special notation for second-order pattern matching: in an
+      applicative pattern of the form :n:`@?@ident @ident__1 … @ident__n`,
+      the variable :token:`ident` matches any complex expression with (possible)
+      dependencies in the variables :n:`@ident__i` and returns a functional term
+      of the form :n:`fun @ident__1 … @ident__n => @term`.
+
+   .. _match_term_context:
+
+   :n:`context {? @ident } [ @cpattern ]`
+      Matches any term with a subterm matching :token:`cpattern`. If there is a match
+      and :n:`@ident` is present, it is assigned the "matched
+      context", i.e. the initial term where the matched subterm is replaced by a
+      hole.  Note that `context`
+      (with very similar syntax) appearing after the `=>` is the :tacn:`context` tactic.
+
+      For :tacn:`match!` and :tacn:`multi_match!`, if the evaluation of the :token:`ltac2_expr`
+      fails, the next matching subterm is tried. If no further subterm matches, the next branch
+      is tried.  Matching subterms are considered from top to bottom and from left to
+      right (with respect to the raw printing obtained by setting the
+      :flag:`Printing All` flag).  :ref:`Example<ltac2_match_term_context_ex>`.
+
+   .. todo There's a more realistic example from @JasonGross here:
+      https://github.com/coq/coq/pull/12103#discussion_r432996954
+
+   :n:`@ltac2_expr`
+      The tactic to apply if the construct matches.  Metavariable values from the pattern
+      match are substituted
+      into :n:`@ltac2_expr` before it's applied.  Note that metavariables are not
+      prefixed with the question mark as they are in :token:`cpattern`.
```

Comment:
```suggestion
      The tactic to apply if the construct matches.  Metavariable values from the pattern
      match are bound as Ltac2 variables in the context of :n:`@ltac2_expr` before it's
      applied.  Note that the names of bound metavariables are not prefixed with question
      marks as they are in :token:`cpattern`.
```
But why is this text here in the first place?  It's redundant with the below
```
Variables from the :n:`@tac2pat1` are statically bound in the body of the branch, to
values of type `constr` for the variables from the :n:`@term` pattern and to a
value of type `Pattern.context` for the variable :n:`@lident`.
```
and less accurate to boot.

Reply:
Check what I did with your suggested change.

I can't make sense of "values of type `constr` for the variables from the :n:`@term` pattern and to a value of type `Pattern.context` for the variable :n:`@lident`."  lident is in the pattern, right?

Reply:
I believe it refers to, e.g., in `match! '(True /\ True) with | context c [and ?x] => ... end`, we'll get `c` bound with type `Pattern.context` and `x` bound with type `constr`.  Does this explain it?

Reply:
That helps.  I think this covers it:

Variables from the :n:`@tac2pat1` are statically bound in the body of the branch. Variables from the :n:`@term` pattern have values of type `constr`. Variables from the :n:`@ident` in the `context` construct have values of type `Pattern.context` (defined in `Pattern.v`).